Projects/AkkaHTTPMicroserviceExample/src/main/scala/AkkaHttpMicroservice.scala

```
@@ -0,0 +1,116 @@
+  implicit val system: ActorSystem
+  implicit def executor: ExecutionContextExecutor
+  implicit val materializer: Materializer
+
+  def config: Config
+  val logger: LoggingAdapter
+
+  lazy val ipApiConnectionFlow: Flow[HttpRequest, HttpResponse, Any] =
+    Http().outgoingConnection(config.getString("services.ip-api.host"), config.getInt("services.ip-api.port"))
+
+  def ipApiRequest(request: HttpRequest): Future[HttpResponse] = Source.single(request).via(ipApiConnectionFlow).runWith(Sink.head)
+
+  def fetchIpInfo(ip: String): Future[Either[String, IpInfo]] = {
+    ipApiRequest(RequestBuilding.Get(s"/json/$ip")).flatMap { response =>
+      response.status match {
+        case OK => Unmarshal(response.entity).to[IpInfo].map(Right(_))
+        case BadRequest => Future.successful(Left(s"$ip: incorrect IP format"))
+        case _ => Unmarshal(response.entity).to[String].flatMap { entity =>
+          val error = s"FreeGeoIP request failed with status code ${response.status} and entity $entity"
+          logger.error(error)
+          Future.failed(new IOException(error))
+        }
+      }
+    }
+  }
+
+  val routes = {
+    logRequestResult("akka-http-microservice") {
+      pathPrefix("ip") {
+        (get & path(Segment)) { ip =>
+          complete {
+            fetchIpInfo(ip).map[ToResponseMarshallable] {
+              case Right(ipInfo) => ipInfo
+              case Left(errorMessage) => BadRequest -> errorMessage
+            }
+          }
+        } ~
+        (post & entity(as[IpPairSummaryRequest])) { ipPairSummaryRequest =>
+          complete {
+            val ip1InfoFuture = fetchIpInfo(ipPairSummaryRequest.ip1)
+            val ip2InfoFuture = fetchIpInfo(ipPairSummaryRequest.ip2)
+            ip1InfoFuture.zip(ip2InfoFuture).map[ToResponseMarshallable] {
+              case (Right(info1), Right(info2)) => IpPairSummary(info1, info2)
+              case (Left(errorMessage), _) => BadRequest -> errorMessage
```

this will eat the error message from the `ip2InfoFuture` if there is any.